azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -7,19 +7,24 @@
 import logging
 import abc
 import six
+import sys
+import time
 import uuid
+import weakref
 from six.moves import queue
+from threading import Timer
 from . import pipeline_events_base
-from . import pipeline_ops_base
-from . import operation_flow
+from . import pipeline_ops_base, pipeline_ops_mqtt
+from .operation_flow import PipelineFlow
 from . import pipeline_thread
 from azure.iot.device.common import handle_exceptions
+from azure.iot.device.common.callable_weak_method import CallableWeakMethod
 
 logger = logging.getLogger(__name__)
 
 
 @six.add_metaclass(abc.ABCMeta)
-class PipelineStage(object):
+class PipelineStage(PipelineFlow):
```

Comment:
>PipelineFlow [](start = 20, length = 12)

This is a very interesting design choice that I had never really considered.

I don't have an opinion on it (yet), but this is some really out of the box thinking here.

Reply:
Really?  OK.  This idea came from you.  You threw this at me before operation_flow.py even got pulled out of pipeline_stages_base.py.  I thinking you were talking about _inheritance_ at the time, and I think about this in terms of _composition_, and maybe that's why you don't recognize it.  

Also, this is the same pattern (sort-of) as you use in SharedClientInstantiationTests.  I've moved to using a similar pattern in Horton, only to a much larger extreme.  My objects for modules and devices are just compositions of smaller objects for features like twins, methods, telemetry, etc.  I do this for tests themselves, for rest wrappers, and for glue.  It gives me huge flexibility for sharing code between clients.   (See here for the best example: [https://github.com/Azure/iot-sdks-e2e-fx/blob/af15021fdf560096f326a563052308aa1d6d3bb0/test-runner/adapters/rest/rest_iothub_apis.py#L327](https://github.com/Azure/iot-sdks-e2e-fx/blob/af15021fdf560096f326a563052308aa1d6d3bb0/test-runner/adapters/rest/rest_iothub_apis.py#L327)) 

---
In reply to: [333702205](https://github.com/Azure/azure-iot-sdk-python/pull/347#discussion_r333702205) [](ancestors = 333702205)

Reply:
Hm. Maybe. I don't recall. 

---
In reply to: [333730282](https://github.com/Azure/azure-iot-sdk-python/pull/347#discussion_r333730282) [](ancestors = 333730282,333702205)